### PR TITLE
Feature/more

### PR DESCRIPTION
**Pull request by AlexandreMarkus**
https://user-images.githubusercontent.com/59664238/109403934-4b60f000-792f-11eb-8c98-05a329c59ca1.mp4

### CODE REVIEW

**Comment by emondhar** [comment]
Woowwwwww!

**Comment by emondhar** [comment]
Bon travail! Est-ce que c’est normal que l’animation lag un peu dans la vidéo?

**Comment by AlexandreMarkus** [comment]
> Bon travail! Est-ce que c’est normal que l’animation lag un peu dans la vidéo?

Le vidéo n'est pas top. Peux-tu lancer l'application voir si ça le fait aussi sur ton téléphone?

**Comment by apomalyn** [comment]
Déjà 👏 👏 👏  Alors l'animation est très très très lente sur téléphone Sinon c'est top!

**Comment by emondhar** [comment]
> > Bon travail! Est-ce que c’est normal que l’animation lag un peu dans la vidéo?
> 
> @emondhar
> Le vidéo n'est pas top. Peux-tu lancer l'application voir si ça le fait aussi sur ton téléphone?

moi j'ai des erreurs: 
image

File: test/managers/settings_manager_test.dart [hunk 1]
```diff
@@ -106,6 +108,86 @@ void main() {
...
 
+    group("ThemeMode - ", () {
+      test("validate default behaviour", () async {
+
+        PreferencesServiceMock.stubGetString(
+            preferencesService as PreferencesServiceMock,
+            PreferencesFlag.theme,
+            toReturn: 'test theme');
+
+        manager.themeMode;
```
Comment (apomalyn) [inline test/managers/settings_manager_test.dart:119]: Why?

File: test/ui/views/more_view_test.dart [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+void main() {
+  group('MoreView - ', () {
+    setUp(() async {
+      setupNavigationServiceMock();
+      setupCacheManagerMock();
+    });
+
+    tearDown(() {});
+
+    group('UI - ', () {
```
Comment (apomalyn) [inline test/ui/views/more_view_test.dart:19]: Add some test to check the navigation when you tap on each list tile :slightly_smiling_face:

File: test/ui/views/settings_view_test.dart [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  group('SettingsView - ', () {
+    setUp(() async {
+      setupNavigationServiceMock();
+      setupCacheManagerMock();
+      setupSettingsManagerMock();
+    });
+
+    tearDown(() {});
+
+    group('UI - ', () {
```
Comment (apomalyn) [inline test/ui/views/settings_view_test.dart:20]: Add some tests to check the opening of popups, and correct behaviour on tap for example the icon of the theme currently selected has multiple states (sunny, night, and a third one if I'm right), that should be tested :slightly_smiling_face: sorry, more tests to write

**Review by apomalyn — CHANGES_REQUESTED** [review]
Good job! I have a few comments and propositions for you :D also, it's important to have analytics on the settings to see what's settings are the most used. Anyway, if you have any questions don't hesitate!! :smile: 

And once again, good job!!!

**Comment by apomalyn** [comment]
Also, need to be tested on iOS.

**Comment by apomalyn** [comment]
I saw a change in the pubspec about firebase_core, check if this can close issue #17

File: lib/ui/views/about_view.dart [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+class _AboutViewState extends State<AboutView> with TickerProviderStateMixin {
+  AnimationController _controller;
+  bool _completed = false;
+
+  @override
+  void initState() {
+    super.initState();
+    _controller =
+        AnimationController(vsync: this, duration: const Duration(seconds: 2));
```
[the remark this answers is not in the document]
Comment (AlexandreMarkus) [reply lib/ui/views/about_view.dart:23]: updated speed, should be better now

File: lib/ui/views/about_view.dart [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    _controller.forward();
+    _controller.addStatusListener((status) {
+      if (status == AnimationStatus.completed) {
+        _completed = true;
+        setState(() {});
```
[the remark this answers is not in the document]
Comment (AlexandreMarkus) [reply lib/ui/views/about_view.dart:28]: done

File: lib/ui/views/about_view.dart [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+                  AppIntl.of(context).more_applets_about_details,
+                  style: const TextStyle(color: Colors.white),
+                ),
+              ),
+              Row(
+                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
+                children: [
+                  IconButton(
+                      icon: Image.asset("assets/images/website_white.png"),
+                      onPressed: () => _launchURL("https://clubapplets.ca/")),
```
[the remark this answers is not in the document]
Comment (AlexandreMarkus) [reply lib/ui/views/about_view.dart:96]: Great! done

File: lib/ui/views/about_view.dart [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+                  IconButton(
+                      icon: Image.asset("assets/images/github_white.png"),
```
[the remark this answers is not in the document]
Comment (AlexandreMarkus) [reply lib/ui/views/about_view.dart:98]: done

**Comment by apomalyn** [comment]
> Also, need to be tested on iOS.

And it's working on iOS!!!! 🎉 🎉

File: lib/ui/views/about_view.dart [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+                      onPressed: () =>
+                          _launchURL("https://github.com/ApplETS")),
```
[the remark this answers is not in the document]
Comment (AlexandreMarkus) [reply lib/ui/views/about_view.dart:100]: done

File: lib/ui/views/about_view.dart [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+                  IconButton(
+                      icon: Image.asset("assets/images/facebook_white.png"),
+                      onPressed: () =>
+                          _launchURL("https://facebook.com/ClubApplETS")),
```
[the remark this answers is not in the document]
Comment (AlexandreMarkus) [reply lib/ui/views/about_view.dart:104]: done

File: lib/ui/views/about_view.dart [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+                  IconButton(
+                      icon: Image.asset("assets/images/twitter_white.png"),
+                      onPressed: () =>
+                          _launchURL("https://twitter.com/ClubApplETS")),
```
[the remark this answers is not in the document]
Comment (AlexandreMarkus) [reply lib/ui/views/about_view.dart:108]: done

File: lib/ui/views/about_view.dart [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+                  IconButton(
+                      icon: Image.asset("assets/images/youtube_white.png"),
+                      onPressed: () => _launchURL(
+                          "https://youtube.com/channel/UCiSzzfW1bVbE_0KcEZO52ew")),
```
[the remark this answers is not in the document]
Comment (AlexandreMarkus) [reply lib/ui/views/about_view.dart:112]: done

File: lib/ui/views/more_view.dart [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+                ListTile(
+                  title: Text(AppIntl.of(context).more_open_source_licenses),
+                  leading: const Icon(Icons.code),
+                  onTap: () => Navigator.of(context).push(
+                    PageRouteBuilder(
+                      pageBuilder: (context, _, __) => AboutDialog(
+                        applicationIcon: const FlutterLogo(),
+                        applicationName:
+                            AppIntl.of(context).more_open_source_licenses,
+                        applicationVersion: '4.x.x',
```
[the remark this answers is not in the document]
Comment (AlexandreMarkus) [reply lib/ui/views/more_view.dart]: great idea! done

**Comment by AlexandreMarkus** [comment]
build is falling need to sort dependencies 

https://dart-lang.github.io/linter/lints/sort_pub_dependencies.html

**Comment by apomalyn** [comment]
Screen Shot 2021-03-04 at 11 23 27 PM

It seems the cache deletion doesn't work...do you have the same issue on android?

File: lib/core/managers/settings_manager.dart [hunk 13]
```diff
@@ -23,6 +25,78 @@ class SettingsManager {
...
+        _locale = null;
+      }
+    });
+    return _locale;
+  }
+
+  /// Set ThemeMode to ThemeMode.dark
+  void setDarkMode() {
+    _themeMode = ThemeMode.dark;
+    _preferencesService.setString(PreferencesFlag.theme, 'dark');
```
[the remark this answers is not in the document]
Comment (AlexandreMarkus) [reply lib/core/managers/settings_manager.dart:66]: found an even better way to do it, let me know what you think about it

File: lib/core/managers/settings_manager.dart [hunk 14]
```diff
@@ -23,6 +25,78 @@ class SettingsManager {
...
+    _preferencesService.setString(PreferencesFlag.theme, 'light');
+    notifyListeners();
+  }
+
+  /// Set ThemeMode to ThemeMode.system
+  void setSystemMode() {
+    _themeMode = ThemeMode.system;
+    _preferencesService.setString(PreferencesFlag.theme, 'system');
+    notifyListeners();
+  }
```
[the remark this answers is not in the document]
Comment (AlexandreMarkus) [reply lib/core/managers/settings_manager.dart]: good idea! done

File: lib/core/viewmodels/contributors_viewmodel.dart [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+
+  /// Used to open a url
+  Future<void> launchURL(String url) async {
+    if (await canLaunch(url)) {
+      await launch(url);
+    } else {
+      showToast(AppIntl.current.error);
+      throw 'Could not launch $url';
+    }
+  }
```
[the remark this answers is not in the document]
Comment (AlexandreMarkus) [reply lib/core/viewmodels/contributors_viewmodel.dart:27]: good idea, done!

File: lib/core/viewmodels/settings_viewmodel.dart [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+  /// Current locale
+  String _currentLocale;
+
+  /// Current theme
+  String _selectedTheme;
+
+  /// Possible theme choices
+  final light = 'light';
+  final dark = 'dark';
+  final system = 'system';
```
[the remark this answers is not in the document]
Comment (AlexandreMarkus) [reply lib/core/viewmodels/settings_viewmodel.dart]: done something else instead

File: lib/core/viewmodels/settings_viewmodel.dart [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+
+  String get selectedTheme {
+    return _selectedTheme == light
+        ? AppIntl.current.light_theme
+        : _selectedTheme == dark
+            ? AppIntl.current.dark_theme
+            : AppIntl.current.system_theme;
```
[the remark this answers is not in the document]
Comment (AlexandreMarkus) [reply lib/core/viewmodels/settings_viewmodel.dart]: done

File: test/managers/settings_manager_test.dart (hunk 1 above)
Comment (AlexandreMarkus) [reply test/managers/settings_manager_test.dart:118]: was not useful

File: .github/workflows/validate_pr.yaml [hunk 18]
```diff
@@ -22,6 +22,5 @@ jobs:
           ENCRYPTED_SIGNETS_API_CERT_PASSWORD: ${{ secrets.ENCRYPTED_SIGNETS_API_CERT_PASSWORD }}
           MAPS_API_KEY: ${{ secrets.MAPS_API_KEY }}
       - run: flutter pub get
-      - run: flutter pub run intl_utils:generate
```
[the remark this answers is not in the document]
Comment (AlexandreMarkus) [reply .github/workflows/validate_pr.yaml:25]: done :)

File: README.md [hunk 19]
```diff
@@ -43,6 +43,13 @@ flutter pub get
 flutter pub run intl_utils:generate
 ```
 
+## Add environment variable for API_KEY
+
+- After cloning the repo, you will have to get the packages and generate the l10n classes. To do that run the following command:
```
[the remark this answers is not in the document]
Comment (AlexandreMarkus) [reply README.md:48]: done !

File: README.md [hunk 20]
```diff
@@ -43,6 +43,13 @@ flutter pub get
 flutter pub run intl_utils:generate
 ```
 
+## Add environment variable for API_KEY
```
[the remark this answers is not in the document]
Comment (AlexandreMarkus) [reply README.md:46]: great idea!

File: README.md [hunk 21]
```diff
@@ -43,6 +43,13 @@ flutter pub get
...
+
+- After cloning the repo, you will have to get the packages and generate the l10n classes. To do that run the following command:
+```
+flutter pub run --dart-define=SOME_VAR=SOME_VALUE
```
[the remark this answers is not in the document]
Comment (AlexandreMarkus) [reply README.md:50]: Yes exactly, but the google maps api key only seems to work as ENV VAR.
The Github api key works fine

File: lib/ui/views/more_view.dart [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+    final List<Widget> aboutBoxChildren = <Widget>[
+      const SizedBox(height: 24),
+      RichText(
+        text: TextSpan(
+          children: <TextSpan>[
+            TextSpan(
+                style: textStyle,
+                text: "Flutter is Google's UI toolkit for building beautiful, "
+                    'natively compiled applications for mobile, web, and desktop '
+                    'from a single codebase. Learn more about Flutter at '),
```
[the remark this answers is not in the document]
Comment (AlexandreMarkus) [reply lib/ui/views/more_view.dart:30]: done

File: test/ui/views/settings_view_test.dart (hunk 3 above)
Comment (AlexandreMarkus) [reply test/ui/views/settings_view_test.dart:20]: done

File: test/ui/views/more_view_test.dart (hunk 2 above)
Comment (AlexandreMarkus) [reply test/ui/views/more_view_test.dart:32]: done

**Comment by apomalyn** [comment]
Could you update the Flutter version needed to build the application (in the requirement section of the README.md) :smile:

File: lib/core/viewmodels/more_viewmodel.dart [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+                  ..onTap =
+                      () => Util().launchURL(AppIntl.current.flutter_website)),
+            TextSpan(style: textStyle, text: '.'),
+          ],
+        ),
+      ),
+    ];
+  }
+
+  /// Create Github issue
```
Comment (apomalyn) [inline lib/core/viewmodels/more_viewmodel.dart:106]:
```suggestion
  /// Create Github issue into the Notre-Dame repository with the labels bugs and the platform used.
  /// The bug report will contain a [file], a description [feedbackText] and also some information about the 
  /// application/device.
```

File: lib/core/viewmodels/more_viewmodel.dart [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+        RepositorySlug.full("ApplETS/Notre-Dame"),
+        IssueRequest(
+            title: 'Issue from ${packageInfo.appName} ',
+            body: " **Describe the issue** \n"
+                "```$feedbackText```\n\n"
+                "**Screenshot** \n"
+                "![screenshot](https://github.com/ApplETS/Notre-Dame-Bug-report/blob/main/${file.path.replaceFirst('storage/emulated/0/Android/data/ca.etsmtl.applets.notredame/files/', '')}?raw=true)\n\n"
+                "**Device Infos** \n"
+                "- **App name:** ${packageInfo.appName} \n"
+                "- **Package name:** ${packageInfo.packageName}  \n"
```
Comment (apomalyn) [inline lib/core/viewmodels/more_viewmodel.dart:120]: Do you think this is really useful? After all the App name is ÉTSMobile and will stay like this, the package name will not change either ^^'

File: lib/core/viewmodels/more_viewmodel.dart [hunk 25]
```diff
@@ -0,0 +1,156 @@
...
+                "- **Version:** ${packageInfo.version} \n"
+                "- **Build number:** ${packageInfo.buildNumber} \n"
+                "- **Platform operating system:** ${Platform.operatingSystem} \n"
+                "- **Platform operating system version:** ${Platform.operatingSystemVersion} \n",
+            labels: ['bug', 'platform: ${Platform.operatingSystem}']));
+  }
+
+  /// Upload picture to Github
```
Comment (apomalyn) [inline lib/core/viewmodels/more_viewmodel.dart:128]:
```suggestion
  /// Upload a file to the ApplETS/Notre-Dame-Bug-report repository
```

File: lib/core/utils/util.dart [hunk 26]
```diff
@@ -0,0 +1,18 @@
+// FLUTTER / DART / THIRD-PARTIES
+import 'package:oktoast/oktoast.dart';
+import 'package:url_launcher/url_launcher.dart';
+
+// OTHER
+import 'package:notredame/generated/l10n.dart';
+
+class Util {
```
Comment (apomalyn) [inline lib/core/utils/util.dart:8]:
```suggestion
class Utils {
```

File: lib/core/utils/util.dart [hunk 27]
```diff
@@ -0,0 +1,18 @@
...
+  /// Used to open a url
+  Future<void> launchURL(String url) async {
+    if (await canLaunch(url)) {
+      await launch(url);
+    } else {
+      showToast(AppIntl.current.error);
+      throw 'Could not launch $url';
+    }
+  }
```
Comment (apomalyn) [inline lib/core/utils/util.dart:17]: Could you make it static? So we would not need to instantiate an instance of `Utils` every time.

File: lib/ui/views/more_view.dart [hunk 28]
```diff
@@ -0,0 +1,117 @@
...
+                  onTap: () => model.navigationService
+                      .pushNamed(RouterPaths.contributors),
+                ),
+                ListTile(
+                  title: Text(AppIntl.of(context).more_open_source_licenses),
+                  leading: const Icon(Icons.code),
+                  onTap: () => Navigator.of(context).push(
+                    PageRouteBuilder(
+                      pageBuilder: (context, _, __) => AboutDialog(
+                        applicationIcon: const FlutterLogo(),
```
Comment (apomalyn) [inline lib/ui/views/more_view.dart:66]: Should be the ApplETS logo I think

File: lib/ui/views/more_view.dart [hunk 29]
```diff
@@ -0,0 +1,117 @@
...
+                      model.navigationService.pushNamed(RouterPaths.settings),
+                ),
+                ListTile(
+                  title: Text(AppIntl.of(context).more_log_out),
+                  leading: const Icon(Icons.logout),
+                  onTap: () => Navigator.of(context).push(
+                    PageRouteBuilder(
+                      pageBuilder: (context, _, __) => AlertDialog(
+                        title: Text(
+                          AppIntl.current.more_log_out,
```
Comment (apomalyn) [inline lib/ui/views/more_view.dart:92]: You have the context here, why using `current` ?

```suggestion
                          AppIntl.of(context).more_log_out,
```

File: lib/ui/views/more_view.dart [hunk 30]
```diff
@@ -0,0 +1,117 @@
...
+                          style: const TextStyle(color: Colors.red),
+                        ),
+                        content: Text(
+                            AppIntl.current.more_prompt_log_out_confirmation),
```
Comment (apomalyn) [inline lib/ui/views/more_view.dart:96]:
```suggestion
                            AppIntl.of(context).more_prompt_log_out_confirmation),
```

File: lib/ui/views/more_view.dart [hunk 31]
```diff
@@ -0,0 +1,117 @@
...
+                        actions: [
+                          TextButton(
+                              onPressed: () async {
+                                model.logout(context);
+                              },
+                              child: Text(AppIntl.current.yes)),
```
Comment (apomalyn) [inline lib/ui/views/more_view.dart:102]:
```suggestion
                              child: Text(AppIntl.of(context).yes)),
```

File: lib/ui/views/more_view.dart [hunk 32]
```diff
@@ -0,0 +1,117 @@
...
+                          TextButton(
+                              onPressed: () => Navigator.of(context).pop(),
+                              child: Text(AppIntl.current.no))
```
Comment (apomalyn) [inline lib/ui/views/more_view.dart:105]:
```suggestion
                              child: Text(AppIntl.of(context).no))
```

File: test/ui/views/more_view_test.dart [hunk 33]
```diff
@@ -0,0 +1,132 @@
...
+          await tester.pumpAndSettle();
+
+          expect(find.text('Yes'), findsOneWidget);
+          expect(find.text('No'), findsOneWidget);
+          expect(find.byType(AlertDialog), findsOneWidget);
+        });
+      });
+
+      group("golden - ", () {
+        testWidgets("default view (no events)", (WidgetTester tester) async {
```
Comment (apomalyn) [inline test/ui/views/more_view_test.dart:120]:
```suggestion
        testWidgets("default view", (WidgetTester tester) async {
```

File: test/ui/views/more_view_test.dart [hunk 34]
```diff
@@ -0,0 +1,132 @@
...
+
+          // Rebuild the widget after the state has changed.
+          await tester.pumpAndSettle();
+
+          expect(find.text('CLOSE'), findsOneWidget);
+          expect(find.text('VIEW LICENSES'), findsOneWidget);
+          expect(find.byType(AboutDialog), findsOneWidget);
+        });
+
+        testWidgets('parameters', (WidgetTester tester) async {
```
Comment (apomalyn) [inline test/ui/views/more_view_test.dart:90]:
```suggestion
        testWidgets('settings', (WidgetTester tester) async {
```

**Review by apomalyn — CHANGES_REQUESTED** [review]

File: lib/core/viewmodels/more_viewmodel.dart (hunk 24 above)
Comment (AlexandreMarkus) [reply lib/core/viewmodels/more_viewmodel.dart:131]: true

File: lib/core/viewmodels/more_viewmodel.dart [hunk 35]
```diff
@@ -0,0 +1,156 @@
...
+                text: AppIntl.current.flutter_website,
+                recognizer: TapGestureRecognizer()
+                  ..onTap =
+                      () => Util().launchURL(AppIntl.current.flutter_website)),
+            TextSpan(style: textStyle, text: '.'),
+          ],
+        ),
+      ),
+    ];
+  }
```
[the remark this answers is not in the document]
Comment (AlexandreMarkus) [reply lib/core/viewmodels/more_viewmodel.dart:113]: ok, changed it back

File: lib/ui/views/more_view.dart (hunk 28 above)
Comment (AlexandreMarkus) [reply lib/ui/views/more_view.dart:66]: yes, good

File: lib/core/utils/util.dart (hunk 27 above)
Comment (AlexandreMarkus) [reply lib/core/utils/util.dart:17]: done

**Review by apomalyn — APPROVED** [review]
LGTM !! :tada:

**Pull request merged**